cache/cache.go

```diff
@@ -0,0 +1,58 @@
+package cache
+
+import (
+	"crypto/sha1"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"github.com/rogpeppe/go-internal/lockedfile"
+)
+
+func Add(key string, addItem func() error) error {
+	if err := os.MkdirAll(key, 0700); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to create directory %s", key))
+	}
+
+	lockFpath := filepath.Join(key, ".lock")
+	lock, err := lockedfile.Create(lockFpath)
```

Comment:
creating directories in Unix is atomic, so you may not need a lockfile.  The worry I would have with lockfiles is that if they fail to cleanup you end up with a deadlock.

Reply:
This will be valid for windows, linux and mac. In windows, we are slowness of steps since docker images takes more time to download. I can check if there is a way to timeout the lock after a default time say 5 mins.

Reply:
We are using golang internal file locking https://pkg.go.dev/cmd/go/internal/lockedfile@go1.20.2. It uses syscall.Flock on unix for locking and as per documentation, the operating system will drop locks for open files when the go command exits. So, we should be safe here from deadlock perspective.

Reply:
If 2 steps run in parallel trying to run the same drone plugin, they both will clone the repo at the same time which can result in inconsistent state. We will need to take a lock.